src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerServices/FunctionPointerOps.cs

```diff
@@ -63,44 +28,16 @@ public static unsafe IntPtr GetGenericMethodFunctionPointer(IntPtr canonFunction
 
             lock (s_genericFunctionPointerDictionary)
             {
-                var key = new GenericMethodDescriptorInfo
-                {
-                    MethodFunctionPointer = canonFunctionPointer,
-                    InstantiationArgument = instantiationArgument
-                };
-
-                uint index = 0;
-                if (!s_genericFunctionPointerDictionary.TryGetValue(key, out index))
+                ref IntPtr descriptor = ref CollectionsMarshal.GetValueRefOrAddDefault(s_genericFunctionPointerDictionary, (canonFunctionPointer, instantiationArgument), out bool exists);
+                if (!exists)
                 {
-                    // Capture new index value
-                    index = s_genericFunctionPointerNextIndex;
+                    descriptor = (IntPtr)NativeMemory.Alloc((uint)sizeof(GenericMethodDescriptor));
```

Comment:
If this throws, we will end up with an entry without a value. You can fix this by checking whether `descriptor` is null instead of whether the entry exists.